### ISSUE TITLE
Database Pod CrashLoopBackOff because wrong permission

### ISSUE THREAD

**Issue by nghnam** [goharbor/harbor-helm#444]
I deployed Harbor database with persistent volume provided by Rook Ceph. 
Version
```
NAME  	REVISION	UPDATED                 	STATUS  	CHART       	APP VERSION	NAMESPACE
harbor	4       	Tue Nov 19 11:29:34 2019	DEPLOYED	harbor-1.2.3	1.9.3      	common  
```

Pod Status
```
harbor-harbor-database-0                       0/1     CrashLoopBackOff   4          2m33s
```

Logs
```
FATAL:  data directory "/var/lib/postgresql/data" has group or world access
DETAIL:  Permissions should be u=rwx (0700).
```

I fixed by adding a `chmod` command in initContainers in file `database/database-ss.yaml` like this
```
      initContainers:
      - args:
        - -c
        - chown -R 999:999 /var/lib/postgresql/data; chmod 700 -R /var/lib/postgresql/data
```

**Comment by miracle152004** [comment]
Could we add this to the official chart.
I have the same problem using local-storage persistent volumes.

**Comment by jbartlet** [comment]
Same here, using 1.9.1 and Rook-Ceph. It actually starts up fine and runs for ~ 1 hour, then the Postgres db fails with the 'wrong permissions, should be 0700' error in @nghnam post. Haven't figured out what triggers the issue after time (maybe a container restart?).

**Comment by adamcavendish** [comment]
@ywk253100 I can confirm that after you switch between nodes the root-level directory permission is not preserved for ceph-rbd claims. In addition, the official postgres documents say that we should specify PGDATA to a subdirectory of /var/lib/postgresql/data if we're using it as a mount point.

https://hub.docker.com/_/postgres

```
PGDATA
This optional variable can be used to define another location - like a subdirectory - for the database files. The default is /var/lib/postgresql/data, but if the data volume you're using is a filesystem mountpoint (like with GCE persistent disks), Postgres initdb recommends a subdirectory (for example /var/lib/postgresql/data/pgdata ) be created to contain the data.
```

**Comment by zolij** [comment]
same problem with ceph rbd backend

**Comment by whmzsu** [comment]
same problem with ceph rbd storage backend.
The problem occur while shutdown the whole cluster and start  it , or shutdown the node where the   database pod run.
May be there is some proble with the database start script.

**Comment by whmzsu** [comment]
I found a value parameter  named "persistence.persistentVolumeClaim.database.subPath“  of this chart,   the default value is blank ,  So I think  we should fill this value with a subdir name  to avoid this problem. I have not test it .  I will give a feedback after my home lab test , maybe this weekend. Anyone who have a test env can check this  value will work or not. If this worked,  we can  make an PR to this project.

**Comment by reasonerjt** [comment]
@adamcavendish I believe the suggestion you mentioned is for keeping the permission of the data volume of the DB?

**Comment by adamcavendish** [comment]
> @adamcavendish I believe the suggestion you mentioned is for keeping the permission of the data volume of the DB?

Yeah, and it is the official recommendation to not mess up with the database data permissions.

**Comment by ywk253100** [comment]
> @ywk253100 I can confirm that after you switch between nodes the root-level directory permission is not preserved for ceph-rbd claims. In addition, the official postgres documents say that we should specify PGDATA to a subdirectory of /var/lib/postgresql/data if we're using it as a mount point.
> 
> https://hub.docker.com/_/postgres
> 
> > PGDATA
> > This optional variable can be used to define another location - like a subdirectory - for the database files. The default is /var/lib/postgresql/data, but **if the data volume you're using is a filesystem mountpoint** (like with GCE persistent disks), Postgres initdb recommends a **subdirectory** (for example /var/lib/postgresql/data/pgdata ) be created to contain the data.

A PR https://github.com/goharbor/harbor-helm/pull/782 with the suggestion from @adamcavendish included has already been merged into the master and the fix will be available in the release 1.6.0.
 
The reason why we don't use the init container solution is that running `chown` needs the root permission,  when the chart is deployed in a restricted pod security policy enabled k8s, running such command isn't permitted.

**Comment by kav** [comment]
Not sure why @ywk253100 mentions chown as that's not the change in question here. The proposed change adds a chmod the chown line has been present for years at this point. chmod is absolutely permitted in that context. If the chown is blocked by security policy that's a completely different issue.

As mentioned in https://github.com/goharbor/harbor/issues/8224#issuecomment-616110121

Changing the chart with: 
```
-        args: ["-c", "chown -R 999:999 /var/lib/postgresql/data"]
+        args: ["-c", "chown -R 999:999 /var/lib/postgresql/data; chmod 0700 /var/lib/postgresql/data"]
```
works fine and is the current solution in place in a number of postgres helm charts.

**Comment by adamcavendish** [comment]
> > @ywk253100 I can confirm that after you switch between nodes the root-level directory permission is not preserved for ceph-rbd claims. In addition, the official postgres documents say that we should specify PGDATA to a subdirectory of /var/lib/postgresql/data if we're using it as a mount point.
> > https://hub.docker.com/_/postgres
> > > PGDATA
> > > This optional variable can be used to define another location - like a subdirectory - for the database files. The default is /var/lib/postgresql/data, but **if the data volume you're using is a filesystem mountpoint** (like with GCE persistent disks), Postgres initdb recommends a **subdirectory** (for example /var/lib/postgresql/data/pgdata ) be created to contain the data.
> 
> A PR #782 with the suggestion from @adamcavendish included has already been merged into the master and the fix will be available in the release 1.6.0.
> 
> The reason why we don't use the init container solution is that running `chown` needs the root permission, when the chart is deployed in a restricted pod security policy enabled k8s, running such command isn't permitted.

The reason that I did not come up with a PR is because of the legacy deployments using helm will break on upgrade. Plz make sure we do not break their apps.

**Comment by ywk253100** [comment]
@adamcavendish Yes, a separated PR will be submit to cover the upgrade part

**Comment by Djiit** [comment]
Hey team,

We just updated our Harbor chart to 1.6.0 and we keep having this error. What's the current workaround ?

**Comment by Djiit** [comment]
FYI The workaround suggest by OP works for us. But I'll have to manually edit the chart if I want to persist it.

**Comment by Djiit** [comment]
Or, put differently, it did break our deployment when we upgrade from 1.5.X to 1.6.0 :/

**Comment by lotherk** [comment]
Upgrading from 1.5.6 to 1.6.0 also did break it for us. We also had to use the workaround to make it work again.

**Comment by lucioveloso** [comment]
+1 and needed to remove setgid bit in the workaround.

**Comment by techgnosis** [comment]
I just ran into this bug on Helm chart 1.6.0. I turned off my homelab then when turning it back on Harbor is unusable due to this permission error

**Comment by ywk253100** [comment]
The permission problem in this issue is already fixed. 

@Djiit @lotherk @lucioveloso @techgnosis Let's talk about the permission issue after upgrading to 1.6.0 in this thread https://github.com/goharbor/harbor-helm/issues/876, thanks

**Comment by vmSalem** [comment]
I am encountering the same problem with harbor 2.2

**Comment by serkirsch** [comment]
I have the same problem with Harbor 2.2 and the Longhorn CSI provider. It runs perfectly on the first start and crashed on a restart.

**Comment by StevenLeiZhang** [comment]
> I have the same problem with Harbor 2.2 and the Longhorn CSI provider. It runs perfectly on the first start and crashed on a restart.
I hit the same error, here is my fix. I just modify the args of "database-ss.yaml", which locates at templates/database/
`        #args: ["-c", "[ -e /var/lib/postgresql/data/postgresql.conf ] && [ ! -d /var/lib/postgresql/data/pgdata ] && mkdir -m 0700 /var/lib/postgresql/data/pgdata && mv /var/lib/postgresql/data/* /var/lib/postgresql/data/pgdata/ || true"]
        args: ["-c", "[ ! -d /var/lib/postgresql/data/pgdata ] && mkdir -m 0700 /var/lib/postgresql/data/pgdata  && [ -e /var/lib/postgresql/data/postgresql.conf ] && mv /var/lib/postgresql/data/* /var/lib/postgresql/data/pgdata/ || chmod 0700 /var/lib/postgresql/data/pgdata && ls -l /var/lib/postgresql/data || true"]
       `
&& ls -l /var/lib/postgresql/data is just used for print debug info, you could remove it.

**Comment by gaboucher** [comment]
On Harbor 2.2, I was having the same issue while restoring the database PVC with Velero. Made sure that Velero was not the problem and based on @StevenLeiZhang answer, you can fix this issue **temporarily** with `kubectl` : 
```
kubectl patch statefulset harbor-harbor-database -p '{"spec": { "template": { "spec": { "initContainers": [{"args":["-c","[ ! -d /var/lib/postgresql/data/pgdata ] && mkdir -m 0700 /var/lib/postgresql/data/pgdata && [ -e /var/lib/postgresql/data/postgresql.conf ] && mv /var/lib/postgresql/data/* /var/lib/postgresql/data/pgdata/ || chmod 0700 /var/lib/postgresql/data/pgdata && ls -l /var/lib/postgresql/data || true"],"command":["/bin/sh"],"image":"goharbor/harbor-db:dev","imagePullPolicy":"IfNotPresent","name":"data-migrator","resources":{},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","volumeMounts":[{"mountPath":"/var/lib/postgresql/data","name":"database-data"}]}] } } } }'
```
Note that if you `helm upgrade` again, you'll have to redo this command.

**Comment by austinorth** [comment]
Same problem here, and looks like many other people are experiencing it. @ywk253100 can we please re-open this issue?

**Comment by Morriz** [comment]
> > @ywk253100 I can confirm that after you switch between nodes the root-level directory permission is not preserved for ceph-rbd claims. In addition, the official postgres documents say that we should specify PGDATA to a subdirectory of /var/lib/postgresql/data if we're using it as a mount point.
> > https://hub.docker.com/_/postgres
> > > PGDATA
> > > This optional variable can be used to define another location - like a subdirectory - for the database files. The default is /var/lib/postgresql/data, but **if the data volume you're using is a filesystem mountpoint** (like with GCE persistent disks), Postgres initdb recommends a **subdirectory** (for example /var/lib/postgresql/data/pgdata ) be created to contain the data.
> 
> A PR #782 with the suggestion from @adamcavendish included has already been merged into the master and the fix will be available in the release 1.6.0.
> 
> The reason why we don't use the init container solution is that running `chown` needs the root permission, when the chart is deployed in a restricted pod security policy enabled k8s, running such command isn't permitted.

Please read this carefully people, before suggesting to change permissions in the chart as it involves root. If you choose to use root you are probably just trying k8s and not using it in production.

**Comment by ren2881971** [comment]
> Update! The parameter "persistence.persistentVolumeClaim.database.subPath“ do not work. The database dir not change after provide a value like “testsubpath”. As @nghnam mentioned, the change of init script do work, I have test it. ######################################################## As @adamcavendish mentioned I found a value parameter named "persistence.persistentVolumeClaim.database.subPath“ of this chart, the default value is blank , So I think we should fill this value with a subdir name to avoid this problem. I have not test it . I will give a feedback after my home lab test , maybe this weekend. Anyone who have a test env can check this value will work or not. If this worked, we can make an PR to this project.

thx, i fill the  "data" value at "persistence.persistentVolumeClaim.database.subPath“ . harbor-database-0   is show running~

**Comment by mumong** [comment]
i meet this issue just now
`init DB, DB version:13
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locales
  COLLATE:  en_US.UTF-8
  CTYPE:    en_US.UTF-8
  MESSAGES: C
  MONETARY: C
  NUMERIC:  C
  TIME:     C
The default text search configuration will be set to "english".

Data page checksums are disabled.

initdb: error: could not create directory "/var/lib/postgresql/data/pgdata": Permission denied
`
i created pv,the valuse is like
`persistence:
  enabled: true
  resourcePolicy: "keep"
  persistentVolumeClaim:
    registry:
      existingClaim: ""
      storageClass: "hostpath"
      subPath: ""
      accessMode: ReadWriteOnce
      size: 100Gi
`
my pv is like
`apiVersion: v1
kind: PersistentVolume
metadata:
  name: harbor-jobservice-avaiable
spec:
  storageClassName: hostpath
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteOnce
  hostPath:
    path: "/opt/kubernetes/data/available_harbor/jobservice"
  nodeAffinity:
    required:
      nodeSelectorTerms:
        - matchExpressions:
            - key: kubernetes.io/hostname
              operator: In
              values:
                - xxx
`
using harbor 2.8.3 helm deployed
one more thing ,when i change my storageClass to silver , it works

**Comment by mumong** [comment]
i meet this error just now
using harbor v2.8.3,helm deployed
error :
![image](https://github.com/goharbor/harbor-helm/assets/91200009/8dd3c504-89a3-4048-a532-097cb705ae44)

pvc
![image](https://github.com/goharbor/harbor-helm/assets/91200009/036dda51-c2ef-4983-921e-9f6486f44bfe)
pv
![image](https://github.com/goharbor/harbor-helm/assets/91200009/a6a6cc0e-828d-46e5-b17a-3d4b9e5bc16f)

i had alreaedy try the solution ,bt not work
![image](https://github.com/goharbor/harbor-helm/assets/91200009/30790d2a-e8aa-4d5e-b533-d034b0d84970)

tips: when i change storageClass to silver.it works
